core/src/main/scala/com/salesforce/op/utils/stages/NameDetectUtils.scala

```diff
@@ -0,0 +1,337 @@
+      (stats.countBelowMaxNumTokens / N) > 0.75,
+      // check that at least 3/4 of the texts are longer than 3 characters
+      (stats.countAboveMinCharLength / N) > 0.75,
+      // check that the standard deviation of the text length is greater than a small number
+      N < 10 || stats.approxMomentsOfTextLength.stddev > 0.05,
+      // check that the number of unique entries is at least 10
+      N < 100 || hllMonoid.sizeOf(stats.approxNumUnique).estimate > 10
+    )
+    checks.forall(identity)
+  }
+
+  private[op] def dictCheck(tokens: Seq[String], dict: NameDictionary): Double = {
+    if (tokens.isEmpty) 0.0 else {
+      tokens.map({ token: String => if (dict contains token) 1 else 0}).sum.toDouble / tokens.length
+    }
+  }
+
+  private[op] def genderDictCheck(nameToCheckGenderOf: String, genderDict: GenderDictionary): String = {
+    genderDict.get(nameToCheckGenderOf).map(
+      probMale => if (probMale >= 0.5) Male else Female
+    ).getOrElse(GenderNA)
+  }
+
+  private[op] def identifyGender(
+    text: String,
+    tokens: Seq[String],
+    strategy: GenderDetectStrategy,
+    genderDict: GenderDictionary
+  ): String = {
+    strategy match {
+      case FindHonorific() =>
+        val matched = tokens filter { NameDetectUtils.AllHonorifics contains }
+        if (matched.length == 1) {
+          if (MaleHonorifics contains matched.head) Male else Female
+        }
+        else GenderNA
+      case ByIndex(index) =>
+        val nameToCheckGenderOf = tokens.lift(index).getOrElse("")
+        genderDictCheck(nameToCheckGenderOf, genderDict)
+      case ByLast() =>
+        val nameToCheckGenderOf = tokens.lastOption.getOrElse("")
+        genderDictCheck(nameToCheckGenderOf, genderDict)
+      case ByRegex(pattern) =>
+        text match {
+          case pattern(matchedGroup) =>
+            val nameToCheckGenderOf = preProcess(Some(matchedGroup)).headOption.getOrElse("")
+            genderDictCheck(nameToCheckGenderOf, genderDict)
+          case _ => GenderNA
+        }
+      case _ =>
+        logError("Unimplemented gender detection strategy found")
+        GenderNA
+    }
+  }
+
+  private[op] def computeGenderResultsByStrategy(
+    text: String,
+    tokens: Seq[String],
+    genderDict: GenderDictionary
+  ): Map[String, GenderStats] = {
+    GenderDetectStrategies map { strategy: GenderDetectStrategy =>
+      val genderResult: String = identifyGender(text, tokens, strategy, genderDict)
+      strategy.toString -> GenderStats(
+        if (genderResult == Male) 1 else 0,
+        if (genderResult == Female) 1 else 0,
+        if (genderResult == GenderNA) 1 else 0
+      )
+    } toMap
+  }
+
+  private[op] def computeResults(
+    input: String,
+    nameDict: Broadcast[NameDictionary],
+    genderDict: Broadcast[GenderDictionary],
+    hll: HyperLogLogMonoid
+  ): NameDetectStats = input match {
+    case "" if $(ignoreNulls) => NameDetectStats.empty
```

Comment:
I think it is a bad practice to handle String values. Better operate on our own `Text` or `Option[String]` and check for emptiness as I pointed our above.

Reply:
Using this thread to discuss all of the other related comments about Text vs. String:
1) The implicit conversions between `T#Value` and `String` are there to make my life easier when I'm porting over to `SmartTextMapVectorizer`, where the values are `String` and not `Text`.
2) As I discovered when erroneously making #449, TMOG does distinguish between `Text.empty` and `Text("")` (where the former is realized with Map types by excluding the relevant key/value from the Map altogether). However, for my application, I would like to treat `Text.empty` and `Text("")` the same way. This is clearer when I just operate over `String`.

With these two points in mind, would you still recommend that I operate over `Text` or `Option[String]`? It's not a big deal, I'm happy to implement it that way.

Reply:
I do this that using container types are beneficial than nullable values such as String. Aa for implicit conversions, they should be avoided as they are difficult to debug and spot in pull request reviews, therefore I would recommend against using them unless absolutely necessary.